### PR TITLE
build: add a script to detect component ID collisions

### PR DESCRIPTION
**Pull request by crisbeto**
As of v16, the framework uses the metadata of a component to generate an ID. If two components have identical metadata, they may end up with the same ID which will log a warning. These changes add a script to detect such cases automatically.

Relates to https://github.com/angular/components/issues/27163.

### CODE REVIEW

**Comment by andrewseguin** [comment]
Why does Angular throw an error here? Isn't our use case valid - we are making a new version of a component.

**Comment by crisbeto** [comment]
tl;dr is that Angular has always had a unique ID for each component. Pre-v16 the ID was based on a counter so components were guaranteed to have a unique ID. When hydration was introduced, the counter approach didn't work anymore since different components might be executed on the server and the client so the ID algorithm was switched to use the metadata instead.

**Comment by andrewseguin** [comment]
Could the ID generation take into account it's filepath or something unique beyond just metadata? It'd be helpful to know they intend to prevent this case otherwise

**Comment by crisbeto** [comment]
It can't since:
1. JIT compilation doesn't know about the file system.
2. This happens in the runtime, not at compilation.

**Comment by andrewseguin** [comment]
Oof...well that's a bummer.

When should we run this script? Seems like it's a proactive thing that we ought to run when we change metadata?

**Comment by devversion** [comment]
Two quick thoughts:

* Should some of this check be performed in the compiler CLI itself? seems like something other libraries would need too.
* Could we just have a runtime test that imports all components and asserts that nothing breaks?

**Comment by crisbeto** [comment]
Things don't necessarily break at the moment, they just log a warning and I wanted to automate catching the warnings instead of having users report it to us.

As for having it in the compiler: I think it's trickier to do it there, because compilations might be spread across different build units running in parallel and things would only conflict once they're on the same page.

**Comment by devversion** [comment]
I thought we could potentially intercept logging and see if there is a matching error/or something, without having to replicate some of the serialization logic. At the same time though- adding all components is likely not super nice either

**Comment by crisbeto** [comment]
Yeah I feel like having it run as a part of the CI against the source code is the simplest without having to maintain even more code.

File: scripts/detect-component-id-collisions.ts [hunk 1]
```diff
@@ -0,0 +1,201 @@
...
+    chalk.red(
+      `\nThe metadata of one of each of these components should be ` +
+        `changed to be slightly different in order to avoid conflicts at runtime.\n`,
+    ),
+  );
+
+  process.exit(1);
+}
+
+/** Checks if a class will conflict with any of the classes that have been check so far. */
```
Comment (devversion) [inline scripts/detect-component-id-collisions.ts:45]:
```suggestion
/** Checks if a class will conflict with any of the classes that have been checked so far. */
```

File: scripts/detect-component-id-collisions.ts [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+function checkClass(node: ts.ClassDeclaration): void {
+  const metadata = getComponentMetadata(node);
+
+  if (metadata) {
+    // Create an ID for the component based on its metadata. This is based on what the framework
+    // does at runtime at https://github.com/angular/angular/blob/main/packages/core/src/render3/definition.ts#L679.
+    // Note that the behavior isn't exactly the same, because the framework uses some fields that
+    // are generated by the compiler based on the component's template.
```
Comment (devversion) [inline scripts/detect-component-id-collisions.ts:53]: Yes, and likely we are overly sensitive here, but it guarantees we are not conflicting.

**Review by devversion — APPROVED** [review]

**Review by andrewseguin — APPROVED** [review]

**Pull request merged**